src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp

```diff
@@ -638,55 +634,130 @@ void serdata_get_keyhash(
+  const struct ddsi_sertype *type,
+  enum ddsi_serdata_kind kind,
+  const char *sample,
+  struct dds_loaned_sample *loan,
+  bool force_serialization)
 {
-  try {
-    auto d = new ddscxx_serdata<T>(typecmn, kind);
+  assert (loan->loan_origin.origin_kind == DDS_LOAN_ORIGIN_KIND_PSMX);
+  const bool key = (kind == SDK_KEY);
+  bool serialize_data = force_serialization || psmx_endpoint_serialization_required(loan->loan_origin.psmx_endpoint);
 
-    // serdata from the loaned sample (when using iceoryx)
-    d->iox_chunk = iox_buffer;
+  const T* sample_in = reinterpret_cast<const T*>(sample);
+  T* sample_out = static_cast<T*>(loan->sample_ptr);
+  struct dds_psmx_metadata *metadata = loan->metadata;
+  ddscxx_serdata<T> *d = new ddscxx_serdata<T>(type, kind);
 
-    // Update the iox subscriber, when constructing the serdata in the case of sample received
-    // from iceoryx
-    if (sub != nullptr) {
-      d->iox_subscriber = sub;
-    }
-
-    // key handling
-    if (typecmn->fixed_size) {
-      const auto& msg = *static_cast<const T*>(d->iox_chunk);
-      d->key_md5_hashed() = to_key(msg, d->key());
-      d->populate_hash();
-    } else {
-      T msg;
-      iceoryx_header_t *iox_header = iceoryx_header_from_chunk(d->iox_chunk);
-      if (deserialize_sample_from_buffer(d->iox_chunk, iox_header->data_size, msg, kind)) {
-        d->key_md5_hashed() = to_key(msg, d->key());
-        d->populate_hash();
-      } else {
+  uint16_t hdr[CDR_HEADER_SIZE*sizeof(uint8_t)/sizeof(uint16_t)] = {DDSI_RTPS_SAMPLE_NATIVE, 0};
+  if (sample_out != sample_in) //the sample we are sending across PSMX has been supplied during the call to dds_write, not a loan from the user
+  {
+    assert (metadata->sample_state == DDS_LOANED_SAMPLE_STATE_UNITIALIZED);
+    if (serialize_data)
+    {
+      bool ser_result = false;
+      if(TopicTraits<T>::allowableEncodings() & DDS_DATA_REPRESENTATION_FLAG_XCDR1)
+          ser_result = serialize_into_impl<T,basic_cdr_stream>(hdr, sample_out, metadata->sample_size, *sample_in, key ? key_mode::unsorted : key_mode::not_key);
+      else if (TopicTraits<T>::allowableEncodings() & DDS_DATA_REPRESENTATION_FLAG_XCDR2)
+          ser_result = serialize_into_impl<T,xcdr_v2_stream>(hdr, sample_out, metadata->sample_size, *sample_in, key ? key_mode::unsorted : key_mode::not_key);
+      if (!ser_result)  //serialization unsuccesful, abort
+      {
         delete d;
-        d = nullptr;
+        return nullptr;
       }
+      metadata->sample_state = (key ? DDS_LOANED_SAMPLE_STATE_SERIALIZED_KEY : DDS_LOANED_SAMPLE_STATE_SERIALIZED_DATA);
     }
+    else
+    {
+      // if you are not serializing, you still need to copy into the loaned sample
+      *sample_out = *sample_in;
+    }
+  }
+
+  if (!serialize_data)
+    metadata->sample_state = (key ? DDS_LOANED_SAMPLE_STATE_RAW_KEY : DDS_LOANED_SAMPLE_STATE_RAW_DATA);
+
+  metadata->cdr_identifier = hdr[0];  //remove endianness?
```

Comment:
I think there's no need to 'remove' any endianness, as the endianness in the header is native

Reply:
You are correct.
I will fix this.